File: leetcode/1-100/5-longest-palindromic-substring/expand_around_center.py
"""LeetCode #5: Longest Palindromic Substring — expand around center.
O(n²) time, O(n) extra space for the indexed char list.

Algorithmic mirror of expand_around_center.kara. Output format matches
line-for-line so the two can be diffed directly.
"""

from __future__ import annotations


def expand(chars: list[str], lo0: int, hi0: int) -> tuple[int, int]:
    # Returns (start, length) of the palindrome centered between lo0 and hi0.
    # Loop invariant matches the Kāra version: on exit, the matched range is
    # chars[lo+1 .. hi-1] inclusive, so start = lo + 1 and length = hi - lo - 1.
    lo, hi = lo0, hi0
    n = len(chars)
    while lo >= 0 and hi < n and chars[lo] == chars[hi]:
        lo -= 1
        hi += 1
    return (lo + 1, hi - lo - 1)


def longest_palindrome(s: str) -> tuple[int, int]:
    chars = list(s)
    n = len(chars)
    best_start = 0
    best_len = 0
    for i in range(n):
        start, length = expand(chars, i, i)
        if length > best_len:
            best_start, best_len = start, length
        start, length = expand(chars, i, i + 1)
        if length > best_len:
            best_start, best_len = start, length
    return (best_start, best_len)


def report(s: str) -> None:
    start, length = longest_palindrome(s)
    print(start)
    print(length)


def main() -> None:
    report("babad")             # expect: 0 3   ("bab")
    report("cbbd")              # expect: 1 2   ("bb")
    report("a")                 # expect: 0 1
    report("ac")                # expect: 0 1
    report("")                  # expect: 0 0
    report("racecar")           # expect: 0 7
    report("abacdfgdcaba")      # expect: 0 3   ("aba")
    report("ababababa")         # expect: 0 9
    report("forgeeksskeegfor")  # expect: 3 10  ("geeksskeeg")
    report("aaa")               # expect: 0 3


if __name__ == "__main__":
    main()

File: leetcode/1-100/5-longest-palindromic-substring/bench/expand_around_center.c
/*
 * LeetCode 5 — expand-around-center Longest Palindromic Substring, C mirror.
 * Algorithmic peer of bench/expand_around_center.{kara,rs,py}. Same input:
 * 5000 'a' chars, K=10, sink = sum of (best_start + best_len) per call.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

static void expand(const char *s, int64_t n, int64_t lo0, int64_t hi0,
                   int64_t *out_lo, int64_t *out_len) {
    int64_t lo = lo0;
    int64_t hi = hi0;
    while (lo >= 0 && hi < n && s[lo] == s[hi]) {
        lo--;
        hi++;
    }
    *out_lo = lo + 1;
    *out_len = hi - lo - 1;
}

static void longest_palindrome(const char *s, int64_t n,
                               int64_t *best_start, int64_t *best_len) {
    int64_t bs = 0;
    int64_t bl = 0;
    for (int64_t i = 0; i < n; i++) {
        int64_t start, length;
        expand(s, n, i, i, &start, &length);
        if (length > bl) {
            bs = start;
            bl = length;
        }
        expand(s, n, i, i + 1, &start, &length);
        if (length > bl) {
            bs = start;
            bl = length;
        }
    }
    *best_start = bs;
    *best_len = bl;
}

int main(void) {
    const int64_t n = 5000;
    char *data = (char *)malloc((size_t)n + 1);
    memset(data, 'a', (size_t)n);
    data[n] = '\0';

    int64_t sum = 0;
    for (int k = 0; k < 10; k++) {
        int64_t start, length;
        longest_palindrome(data, n, &start, &length);
        sum += start + length;
    }
    printf("%lld\n", (long long)sum);
    free(data);
    return 0;
}
